obs-studio-server/source/nodeobs_settings.cpp

```diff
@@ -4310,17 +4316,145 @@ void getDevices(const char *source_id, const char *property_name, std::vector<ip
+	HRESULT hr;
+	DWORD count = 0;
+	ComPtr<IPropertyBag> propertyData;
+
+	hr = CoCreateInstance(CLSID_SystemDeviceEnum, NULL, CLSCTX_INPROC_SERVER, IID_ICreateDevEnum, (void **)&deviceEnum);
+	if (FAILED(hr)) {
+		blog(LOG_ERROR, "Could not create ICreateDeviceEnum");
+		return;
+	}
+
+	hr = deviceEnum->CreateClassEnumerator(type, &enumMoniker, 0);
+	if (FAILED(hr)) {
+		blog(LOG_ERROR, "CreateClassEnumerator failed");
+		return;
+	}
+
+	uint32_t nbDevices = 0;
+	if (hr == S_OK) {
+		VARIANT deviceName, devicePath;
+		deviceName.vt = VT_BSTR;
+		devicePath.vt = VT_BSTR;
+		devicePath.bstrVal = NULL;
+		while (enumMoniker->Next(1, &deviceInfo, &count) == S_OK) {
+			hr = deviceInfo->BindToStorage(0, 0, IID_IPropertyBag, (void **)&propertyData);
+			if (FAILED(hr))
+				continue;
+			hr = propertyData->Read(L"FriendlyName", &deviceName, NULL);
+			if (FAILED(hr))
+				continue;
+
+			char *utf8Name = NULL;
+			os_wcs_to_utf8_ptr(deviceName.bstrVal, 0, &utf8Name);
+			std::string deviceId = utf8Name;
+			deviceId += ':'; // Not a bug, dshow expects it as a separator
+
+			hr = propertyData->Read(L"DevicePath", &devicePath, NULL);
+			if (SUCCEEDED(hr)) {
+				char *utf8Path = NULL;
+				os_wcs_to_utf8_ptr(devicePath.bstrVal, 0, &utf8Path);
+				deviceId += utf8Path;
+			}
+
+			nbDevices++;
+			rval.push_back(ipc::value(utf8Name));
+			rval.push_back(ipc::value(deviceId));
+		}
+	}
+	rval[1] = ipc::value(nbDevices);
+}
+
+std::string GetDeviceName(IMMDevice *device)
+{
+	if (!device) {
+		return "";
+	}
+	std::string device_name;
+	ComPtr<IPropertyStore> store;
+	HRESULT res;
+
+	if (SUCCEEDED(device->OpenPropertyStore(STGM_READ, store.Assign()))) {
+		PROPVARIANT nameVar;
+
+		PropVariantInit(&nameVar);
+		res = store->GetValue(PKEY_Device_FriendlyName, &nameVar);
+
+		if (SUCCEEDED(res) && nameVar.pwszVal && *nameVar.pwszVal) {
+			size_t len = wcslen(nameVar.pwszVal);
+			size_t size;
+
+			size = os_wcs_to_utf8(nameVar.pwszVal, len, nullptr, 0);
+			device_name.resize(size);
+			os_wcs_to_utf8(nameVar.pwszVal, len, &device_name[0], size);
+		}
+	}
+
+	return device_name;
+}
+
+void enumAudioOutputDevices(std::vector<ipc::value> &rval)
+{
+	ComPtr<IMMDeviceEnumerator> enumerator;
+	ComPtr<IMMDeviceCollection> collection;
+	UINT count;
+	HRESULT res;
+
+	res = CoCreateInstance(__uuidof(MMDeviceEnumerator), NULL, CLSCTX_ALL, __uuidof(IMMDeviceEnumerator), (void **)enumerator.Assign());
+	if (FAILED(res))
+		blog(LOG_ERROR, "Failed to create enumerator");
+
+	res = enumerator->EnumAudioEndpoints(eRender, DEVICE_STATE_ACTIVE, collection.Assign());
+	if (FAILED(res))
+		blog(LOG_ERROR, "Failed to enumerate devices");
+
+	res = collection->GetCount(&count);
+	if (FAILED(res))
+		blog(LOG_ERROR, "Failed to get device count");
+
+	rval[1] = ipc::value((uint32_t)count);
+
+	for (UINT i = 0; i < count; i++) {
+		ComPtr<IMMDevice> device;
+		CoTaskMemPtr<WCHAR> w_id;
+
+		res = collection->Item(i, device.Assign());
+		if (FAILED(res))
```

Comment:
if we skip some devices. "count" put inside "rval[1]" became incorrect.

Reply:
oops, good catch, you're right. Fixing

Reply:
fixed